Add support for custom scalars

jgosmann opened the pull request:
## Description

This adds support for custom GraphQL scalars. The usage is a bit different from what was proposed in #114: the serialize/parse functions are passed as argument to the decorator instead of being defined on the type itself. This allows to use an existing type without any changes to the type itself as custom scalar.

Usage examples can be found in the docstring of the `strawberry.scalar` decorator and the accompanying tests.

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* Closes #114.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project. (as far as I can tell)
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

Add support for custom GraphQL scalars.

<!-- action-check: release-file -->

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -0,0 +1,56 @@
+    >>> strawberry.scalar(
+    >>>     datetime.date,
+    >>>     serialize=lambda value: value.isoformat(),
+    >>>     parse_value=datetime.parse_date)
```
```suggestion
    >>>     parse_value=datetime.parse_date
    >>> )
```

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -0,0 +1,56 @@
+    >>> Base64Encoded = strawberry.scalar(
+    >>>     NewType("Base64Encoded", bytes),
+    >>>     serialize=base64.b64encode,
+    >>>     parse_value=base64.b64decode)
```
```suggestion
    >>>     parse_value=base64.b64decode
    >>> )
```

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -0,0 +1,56 @@
+
+    >>> @strawberry.scalar(
+    >>>     serialize=lambda value: ",".join(value.items),
+    >>>     parse_value=lambda value: CustomList(value.split(",")))
```
```suggestion
    >>>     parse_value=lambda value: CustomList(value.split(","))
    >>> )
```

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -0,0 +1,56 @@
+    return cls
+
+
+def scalar(cls=None, *, description=None, serialize, parse_value, parse_literal=None):
```
what if we add a default serializer and parse value? This could make it easy to implement scalars that are just aliases of normal types, like:

```
Currency = strawberry.scalar(int, description="Currency in cents.")
Code = strawberry.scalar(str, description="Conference code.")
```

jgosmann replied:
Note that the could would need to actually look like this (if you don't want to overwrite the handling of all ints and strings):
```python
Currency = strawberry.scalar(NewType("Currency", int), description="Currency in cents.")
Code = strawberry.scalar(NewType("Code", str), description="Conference code.")
```

I added a commit implementing this.

patrick91 commented on tests/test_custom_scalar.py:
```diff
@@ -0,0 +1,68 @@
+    pass
+
+
+@pytest.mark.asyncio
```
is there any particular reason why this test is async? I think we can use graphql sync here ;)

jgosmann replied:
The `execute` function is defined as `async`. I could wrap the call in an `asyncio.run`. Not sure if that is better?

patrick91 replied:
There's also graphql sync we could use, maybe we should make execute sync function 😊 
https://github.com/strawberry-graphql/strawberry/blob/master/tests/test_schema.py#L9

jgosmann replied:
I changed the test to use `graphql_sync`.

patrick91 commented on tests/test_custom_scalar.py:
```diff
@@ -0,0 +1,68 @@
+    schema = strawberry.Schema(Query)
+
+    encoded = base64.b64encode(b"decoded").decode("ascii")
+    result = await execute(schema, f'{{ decodeBase64(encoded: "{encoded}") }}')
```
you can also pass variables to execute, so you don't have to use the double brackets example :)

jgosmann replied:
True, I made a commit to pass it as variable. It's not really any shorter, but maybe more readable? Let me know what you think.

jgosmann commented:
Added custom scalars for `Date`, `DateTime`, and `Time`. They can't go into `scalars.py` because that leads to a cyclic import (because `strawberry.scalar` needs to be imported there, which imports `type_converter.py` which again imports `scalars.py`).

jgosmann commented:
Seems like `aniso8601` isn't included in the dependencies? Should it be added or should we rely on the Python `fromisoformat` methods (which apparently cannot parse arbitrary ISO 8601 strings).

patrick91 commented on strawberry/custom_scalar.py:
```diff
@@ -0,0 +1,69 @@
+    if serialize is None:
+
+        def serialize(x):
+            return x
```
do we have to create a function everytime? we could use a generic one, no?

jgosmann replied:
Yes, I changed it.

patrick91 commented on tests/test_custom_scalar.py:
```diff
@@ -0,0 +1,90 @@
+    query = """query decode($encoded: Base64Encoded!) {
+        decodeBase64(encoded: $encoded)
+    }"""
+    result = await execute(schema, query, variable_values={"encoded": encoded})
```
👍

patrick91 commented on strawberry/datetime_scalar.py:
```diff
@@ -0,0 +1,29 @@
+    description="Time (isoformat)",
+    serialize=_serialize_isoformatted,
+    parse_value=aniso8601.parse_time,
+)
```
can we add some tests for these? :) also I'd put them inside `strawberry/types/datetime.py`

patrick91 replied:
do you think we need to add parse_literal as well? https://github.com/graphql-python/graphene/blob/master/graphene/types/datetime.py

jgosmann replied:
I'll add tests later this week.

I'm not sure about `parse_literal`. It is unclear to me under which conditions this method gets invoked (and I find neither the graphql-core nor the Graphene documentation very helpful on this). [There also seems to be a default implemention](https://github.com/graphql-python/graphql-core-next/blob/712ef8435bbc3066e73b69da96b09c13de945080/src/graphql/type/definition.py#L394), but not sure if that one is appropriate.

jgosmann replied:
So I think I finally understand `parse_literal` (or at least better than before): the GraphQL query is parsed into an AST, then `parse_literal` is used to convert the node representing the value of a scalar into a native Python object, and then `parse_value` converts that Python object to the final object that we want for the scalar.

The date/time objects will be serialized and deserialized as simple strings. Thus, the default `parse_literal` converting the `StringValueNode` from the AST to a Python string is sufficient because we then take that string in `parse_value` to convert it into the proper object. I also added a test verifying the parsing of literal date/time objects.

Looking at [Graphene](https://github.com/graphql-python/graphene/blob/master/graphene/types/scalars.py) the `parse_literal` could be used to ensure that the AST node if of the expected type. But the `parse_value` would also fail (with a reasonable exception) if we do not get a `StringValueNode`.

patrick91 replied:
That makes sense! Thanks for spending time on this!

jgosmann commented:
I update the PR.

* I added `aniso8601` as an optional dependency. This dependency is not installed on the CI system (because it is optional). I need to figure out how to fix that.
* Also, the pytest-mypy-plugins 1.1 version is breaking the tests for me, so I fixed the version to `~1.0` (I guess fixing the issue would be preferrable, but I'd consider that out of the scope of this PR).

The question regarding `parse_literal` is still open.

patrick91 commented:
@jgosmann turns out that it is used when parsing arguments passed directly, instead of using variables, see:

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/667029/67642187-2b678980-f901-11e9-9aae-bf5e5967d090.png">

patrick91 commented on poetry.lock:
```diff
@@ -645,16 +652,18 @@ version = "0.6.0"
 more-itertools = "*"
 
 [extras]
+datetime = ["aniso8601"]
```
maybe we should add this as default dependency :)

jgosmann replied:
I was thinking that someone who doesn't need the DateTime type, might not want to install this dependency. But maybe it is common enough to warrant making this a hard dependency.

patrick91 replied:
@jgosmann I'd add this as default, graphene does that too (with another library) :)

jgosmann replied:
Did that.

patrick91 commented on pyproject.toml:
```diff
@@ -39,10 +40,11 @@ pytest-cov = "^2.8"
 flake8-bugbear = "^19.8"
 pytest-django = {version = "^3.5", optional = true}
-pytest-mypy-plugins = "^1.0"
+pytest-mypy-plugins = "~1.0"
```
is there any reason for this change?

jgosmann replied:
Yes, see [this comment](https://github.com/strawberry-graphql/strawberry/pull/191#issuecomment-546462128).

jgosmann replied:
I merged master into this branch and there it has been bumped to `^1.1` and is working. :)

jgosmann commented:
I'll take a look at `parse_literal` later this week. Thanks, for figuring that out!
